File: Data Structures & Algorithms/validate-parentheses/submission-4.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> bracketStack;
        char currBracket;

        for (int i = 0; i < s.size(); i++)
        {
            currBracket = s[i];

            switch(currBracket)
            {
                case ')':
                    if (bracketStack.size() == 0 || bracketStack.top() != '(')
                    {
                        return false;
                    }
                    bracketStack.pop();
                    break;
                case ']':
                    if (bracketStack.size() == 0 || bracketStack.top() != '[')
                    {
                        return false;
                    }
                    bracketStack.pop();
                    break;
                case '}':
                    if (bracketStack.size() == 0 || bracketStack.top() != '{')
                    {
                        return false;
                    }
                    bracketStack.pop();
                    break;
                default:
                    bracketStack.push(currBracket);
            }
        }

        if (bracketStack.size() != 0)
        {
            return false;
        }
        
        return true;
    }
};

File: Data Structures & Algorithms/car-fleet/submission-4.py
class Solution:
    def carFleet(self, target: int, position: List[int], speed: List[int]) -> int:
        def getTimeToTarget(pos, sp):
            return (target - pos) / sp
        
        cars = sorted(zip(position, speed), reverse=True)
        #print(cars)
        
        stack = []

        # From car closest to target to furthest
        for pos, sp in cars:
            stack.append(getTimeToTarget(pos, sp))
            #print(stack)

            # If car in next (lower) position gets to target faster, then its part of the same fleet
            if len(stack) > 1 and stack[-1] <= stack[-2]:
                #print("popped")
                stack.pop()

        return len(stack)

